base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -0,0 +1,259 @@
+    db: Arc<T>,
+}
+
+impl<T> BlockChainDatabase<T>
+where T: BlockchainBackend
+{
+    /// Reads the blockchain metadata (block height etc) from the underlying backend and returns it.
+    fn read_metadata(db: &T) -> Result<ChainMetadata, ChainStorageError> {
+        let height = match db.get(&DbKey::Metadata(MetadataKey::ChainHeight)) {
+            Ok(Some(DbValue::Metadata(MetadataValue::ChainHeight(v)))) => v,
+            Ok(None) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "The chain height entry is not present in the database. Assuming the database is empty."
+                );
+                0
+            },
+            Err(e) => return log_error(DbKey::Metadata(MetadataKey::ChainHeight), e),
+            Ok(Some(other)) => return unexpected_result(DbKey::Metadata(MetadataKey::ChainHeight), other),
+        };
+
+        let work = match db.get(&DbKey::Metadata(MetadataKey::AccumulatedWork)) {
+            Ok(Some(DbValue::Metadata(MetadataValue::AccumulatedWork(v)))) => v,
+            Ok(None) => {
+                warn!(
+                    target: LOG_TARGET,
+                    "The accumulated work entry is not present in the database. Assuming the database is empty."
+                );
+                0
+            },
+            Err(e) => return log_error(DbKey::Metadata(MetadataKey::AccumulatedWork), e),
+            Ok(Some(other)) => return unexpected_result(DbKey::Metadata(MetadataKey::AccumulatedWork), other),
+        };
+
+        Ok(ChainMetadata {
+            height_of_longest_chain: height,
+            greatest_accumulated_work: work,
+        })
+    }
+
+    /// Creates a new `BlockchainDatabase` using the provided backend.
+    pub fn new(db: T) -> Result<Self, ChainStorageError> {
+        let metadata = Self::read_metadata(&db)?;
+        Ok(BlockChainDatabase {
+            metadata: Arc::new(RwLock::new(metadata)),
+            db: Arc::new(db),
+        })
+    }
+
+    /// If a call to any metadata function fails, you can try and force a re-sync with this function. If the RWLock
+    /// is poisoned because a write attempt failed, this function will replace the old lock with a new one with data
+    /// freshly read from the underlying database. If this still fails, there's probably something badly wrong and
+    /// the thread will panic.
+    pub fn try_recover_metadata(&mut self) -> bool {
+        if !self.metadata.is_poisoned() {
+            // metadata is fine. Nothing to do here
+            return false;
+        }
+        match BlockChainDatabase::read_metadata(self.db.as_ref()) {
+            Ok(data) => {
+                self.metadata = Arc::new(RwLock::new(data));
+                true
+            },
+            Err(e) => {
+                error!(
+                    target: LOG_TARGET,
+                    "Could not read metadata from database. {}. We're going to panic here. Perhaps restarting will \
```

Comment:
lulz. Maybe add a TODO here to come back and review this later. I think the caller should decide whether to panic or not

Reply:
At this point, a thread has panicked while writing to this table, and now it's failed with a new mutex trying to read.

I'd be panicky at this point :-)

Reply:
I guess we could make it return `OMFGThingsAreBorkedYouShouldShutDown` or similar